### PR TITLE
Ecrecover point at infinity

### PR DESCRIPTION
**Pull request by kclowes**
### What was wrong?

`ecdsa_recover` returned a public address even if the Jacobian transformation returned an invalid point (the point at infinity). This behavior was inconsistent with other clients. See: https://github.com/ethereum/tests/issues/941#

Once this is merged and released, we'll need to update `eth-account` and `py-evm`.

### How was it fixed?

Replaced points in the Jacobian files that were meant to represent the points at infinity with the more commonly used representations. And now the `ecdsa_recovery` method checks to see if `Q` is a point at infinity and raises an error to be more consistent with Geth's ecrecover. 

I went back and forth deciding whether to tack on an `is_infinity` flag to the point at infinity representation, but ultimately decided to go with the simple solution for now. I'm open to adding an `is_infinity` flag though if anyone else feels strongly. 

#### Cute Animal Picture

Cute animal picture

### CODE REVIEW

**Comment by kclowes** [comment]
👋 @ChihChengLiang! I tagged you because Piper said you know the math behind this really well. If you could please take a look to make sure these changes make sense, that would be great!

File: tests/backends/test_native_backend_against_coincurve.py [hunk 1]
```diff
@@ -31,12 +31,12 @@
...
 
 
-@pytest.fixture
+@pytest.fixture(scope="module")
 def native_key_api():
     return KeyAPI(backend=NativeECCBackend())
 
 
-@pytest.fixture
+@pytest.fixture(scope="module")
```
Comment (carver) [inline tests/backends/test_native_backend_against_coincurve.py]: Why this change?

File: eth_keys/constants.py [hunk 2]
```diff
@@ -14,6 +14,7 @@
 
 
 #
-# Jacobian Point at infinity
+# Jacobian Representation of the Point at Infinity
+# See: https://crypto.stackexchange.com/questions/41468/point-at-infinity-for-jacobian-coordinates/41473  # noqa: E501
```
Comment (carver) [inline eth_keys/constants.py]: Woah, so I picked up about 10% of what was in that writeup. Definitely interested in someone else taking a look to confirm.

File: tests/backends/test_native_backend_against_coincurve.py [hunk 3]
```diff
@@ -84,33 +84,53 @@ def test_signing_is_equal(private_key_bytes,
...
+def test_native_to_coincurve_recover(private_key_bytes,
+                                     message_hash,
+                                     native_key_api,
+                                     coincurve_key_api):
+
+    native_private_key = native_key_api.PrivateKey(private_key_bytes)
+    native_public_key = native_private_key.public_key
+    native_signature = native_key_api.ecdsa_sign(message_hash, native_private_key)
+
+    coincurve_public_key = coincurve_key_api.ecdsa_recover(message_hash, native_signature)
```
Comment (carver) [inline tests/backends/test_native_backend_against_coincurve.py]: Hm, shouldn't the coincurve backend also complain about the point at infinity? This makes it look like we only fixed the bug when using the native back end.

File: tests/backends/test_native_backend_against_coincurve.py [hunk 4]
```diff
@@ -84,33 +84,53 @@ def test_signing_is_equal(private_key_bytes,
...
+    # The hypothesis message_hash_st shrinks the message hash,
+    # so it tests the zero byte message hash every time.
+    # If message_hash == b'\x00' * 32, the corresponding Jacobian point
+    # on the curve is the point at infinity.
+    # The native backend raises a BadSignature error on ecdsa_recover
+    # for the point at infinity,
+    # whereas the coincurve library returns an address.
+    # This test does introduce some flakiness on the small chance
+    # that hypothesis generates another message hash that corresponds to the
+    # Jacobian point at infinity
```
Comment (carver) [inline tests/backends/test_native_backend_against_coincurve.py]: Hm, that seems important. I wonder if we can track down another message hash that also corresponds to the Jacobian point at infinity.

File: eth_keys/backends/native/ecdsa.py [hunk 5]
```diff
@@ -159,6 +160,10 @@ def ecdsa_raw_recover(msg_hash: bytes,
     XY = jacobian_multiply((x, y, 1), s)
     Qr = jacobian_add(Gz, XY)
     Q = jacobian_multiply(Qr, inv(r, N))
+
+    if Q == POINT_AT_INFINITY:
+        raise BadSignature("InvalidSignature")
```
Comment (carver) [inline eth_keys/backends/native/ecdsa.py]: So I guess this presumes that we correctly catch all the infinity points before they return, to normalize them to the new constant version? (because otherwise, we'd need to test in a broader way, to find out if it's an infinite point)

**Review by carver — COMMENTED** [review]
Yeah, I definitely don't understand what's going on here well enough to give a 👍🏻 . The thing that stood out the most to me was that this only seems to fix the problem if we are using the native backend, but we should definitely make sure this fixes the issue for all backends.

**Comment by kevaundray** [comment]
I think two bugs are introduced with this PR, below I put the rationale.

**TLDR**

- The distinguished point was changed to be (1,1,0), but the code is not properly checking for this point anymore when doing Jacobean arithmetic. 
- The new distinguished point is now the point at infinity which has many representations and so checking for just one of them would not be enough. All representations are (t^2, t^3, 0) where t != 0
 
_(Note: I use the term point at infinity, quite loosely below and it should instead be distinguished point since before this PR, those points were not the points at infinity)_

**Assumptions**

1) The points at infinity, act as the identity element in the group. This means that given a point P, P + POINT_AT_INFINITY = P

2) Previously, the code checked for the point at infinity by checking the second component of the Jacobian Point to be 0

**Evidence for second assumption**

See:

- <https://github.com/ethereum/eth-keys/blob/master/eth_keys/backends/native/jacobian.py#L28>
- <https://github.com/ethereum/eth-keys/blob/master/eth_keys/backends/native/jacobian.py#L41-L44>
- <https://github.com/ethereum/eth-keys/blob/master/eth_keys/backends/native/jacobian.py#L71>

If you check the jacobian_add method (2nd link), where we are adding two points `p` and `q`  , the code checks the second component of `p` using `if not p[1]` and on this condition, it returns `q`.  According to the first assumption I made, this is only true, if `p` is the point identity element. So that if statement is a way of checking for the point at infinity / identity element.

- If we also check jacobian_multipy method (3rd link), where we are doing a scalar multiplication of the point `a` with the scalar `n` , the code first checks the second component of `a` using `if a[1] == 0` , if this is the case then it returns the identity point. This code is essentially saying, if the point is the identity point/ point at infinity, just return it because doing a scalar multiplication with the identity point/point at infinity always returns the identity point. 

So both if statements check the second component of the jacobian point, and if it is zero, it says "this is the point at infinity/identity point" judging by what it does on those conditions.

**Pre PR**

Before the PR was merged, since the points at infinity were represented using  either (0,0,0) or (0,0,1), the check above would work. 

_Note this check implicitly assumes that any point whose second component is 0, is the identity point, I haven't checked any further into assumption, I just assumed it held true._

**Bug**

If you agree with the above rationale, then I think the first bug was introduced because you now represent the point at infinity using the constant (1,1,0) .

See here: <https://github.com/kclowes/eth-keys/blob/ecrecover-point-at-infinity/eth_keys/constants.py#L20>

But the logic to check for the point at infinity is still to check for the second component, which is not zero using that constant.

The second bug is because using the representation that you chose, the points at infinity is now a set of points (t^2, t^3, 0) where t !=0. This is in the stack overflow post accompanied with this PR above. This means that the `if point == POINT_AT_INFINITY` check would not be enough, since it is only checking for one particular representation of the point at infinity, when t = 1, but AFAICT there are no guarantees that this will be the representation that you will get after doing a sequence of point arithmetics. 

**Testing for bug**

- A test which should fail the CI is to add the point at infinity to itself multiple times, then check that the result is the point at infinity.

Something like the below code

```py
result = POINT_AT_INFINITY
for i in range(0,100):
 result = jacobian_add(result, POINT_AT_INFINITY)
assert result == POINT_AT_INFINITY
```

- Another test you can do is to add the point at infinity to a point `P` and check that the result is `P` , where `P` is not the point at infinity.

If I understand your changes, then the two tests should fail.

**Suggestion**

The code was failing because the ecdsa recover was not checking for the points at infinity IIUC. If we assume that the library is correct in just checking for the second component, then you would only need to check this too in the ecdsa_recover method. You could also explicitly check for the two internal representations being used.

I'd probably add a method called `is_identity(p: Tuple[int, int, int]) -> bool` that conveys the fact that you are checking for the identity point and abstracts away the implementation details.

**Comment by ChihChengLiang** [comment]
@kclowes Sorry I didn't finish my review in time. I think @kevaundray summarized the issue thoroughly.

File: tests/backends/test_native_backend_against_coincurve.py (hunk 1 above)
Comment (kclowes) [reply tests/backends/test_native_backend_against_coincurve.py]: There is a hypothesis deprecation warning that says: 
```
<test name> uses the coincurve_key_api fixture, which is reset between function calls 
but not between test cases generated by '@given(...)'.  
You can change it to a module- or session-scoped fixture if it is safe to reuse; 
if not we recommend using a context manager inside your test function.  
See https://docs.pytest.org/en/latest/fixture.html#sharing-test-data for details on fixture scope.
```
So I decided to make that change in this PR. Happy to pull it out into it's own PR if you have a preference.

File: tests/backends/test_native_backend_against_coincurve.py (hunk 3 above)
Comment (kclowes) [reply tests/backends/test_native_backend_against_coincurve.py]: You're right! The most recent changes should work with both backends.

File: tests/backends/test_native_backend_against_coincurve.py (hunk 4 above)
Comment (kclowes) [reply tests/backends/test_native_backend_against_coincurve.py]: Yeah, probably a good idea to track down another one for some more testing. I'll see if I can find one!

File: eth_keys/backends/native/ecdsa.py (hunk 5 above)
Comment (kclowes) [reply eth_keys/backends/native/ecdsa.py]: Yep, that is the presumption. I think this becomes clearer with the most recent changes.

**Comment by kclowes** [comment]
Thanks for the review @kevaundray! I think I fixed the bugs you pointed out, but would appreciate if you could take another look.

> If we assume that the library is correct in just checking for the second component, then you would only need to check this too in the ecdsa_recover method. You could also explicitly check for the two internal representations being used.

Is there a way that seems more correct? If the y-coordinate of point Q [from here](https://github.com/ethereum/eth-keys/pull/76/files#diff-526547dbdfcae48a67cc41fdb4664a5f98a4fe3bd4e3645d5dfa1f174eba6303R162) is 0, does that always correspond to an identity point?


@carver - would appreciate another look as well when you have a few!

File: eth_keys/constants.py [hunk 6]
```diff
@@ -14,7 +14,6 @@
 
 
 #
-# Jacobian Representation of the Point at Infinity
-# See: https://crypto.stackexchange.com/questions/41468/point-at-infinity-for-jacobian-coordinates/41473  # noqa: E501
+# Internal representations of Jacobian identity points
 #
-POINT_AT_INFINITY = (1, 1, 0)
+IDENTITY_POINTS = ((0, 0, 0), (0, 0, 1))
```
Comment (carver) [inline eth_keys/constants.py:19]: Should `(1, 1, 0)` also be in this list? What other points should be?

Comment (kclowes) [reply eth_keys/constants.py:19]: I don't _think_ so, but would like to double check that with @kevaundray. IIUC, based on the feedback he gave above, we should have all identity points covered if the y-coordinate is zero for this implementation. In which case, maybe it makes more sense to have the `is_identity` function check that the y-coordinate is zero rather than matching all coordinates.

Comment (kevaundray) [reply eth_keys/constants.py:19]: yeah I believe only those two points are needed, since the code can only produce those two distinguished points as the identity points

**Review by Besiktas19033 — APPROVED** [review]

**Review by carver — APPROVED** [review]
Cool, I'm mostly piggy-backing on the expertise of other folks here when I say :shipit:
